### PR TITLE
fix(txn): validate client id by public key

### PR DESCRIPTION
**Pull request by cnlangzi**
## Fixes
- validate client id by public key before cache client object.
## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- gosdk: https://github.com/0chain/gosdk/pull/746

### CODE REVIEW

File: code/go/0chain.net/core/encryption/bls0chain.go [hunk 1]
```diff
@@ -239,6 +239,17 @@ func (b0 *BLS0ChainScheme) GetSignature(signature string) (*bls.Sign, error) {
...
 
+// GetClientID - get client id by public key
+func (b0 *BLS0ChainScheme) GetClientID(publicKey string) (string, error) {
+	pk := &bls.PublicKey{}
+	err := pk.DeserializeHexStr(publicKey)
+	if err != nil {
+		return "", err
+	}
+
+	return encryption.Hash(pk.Serialize()), nil
```
Comment (peterlimg) [inline code/go/0chain.net/core/encryption/bls0chain.go]: ClientID is `encryption.Hash(publicKeyBytes)`, we don't need to implement it on `Scheme`, call https://github.com/0chain/0chain/blob/3ab1f87a4fa6eb0df2cc36f5681fbc458d3bf3b7/code/go/0chain.net/core/encryption/signature_scheme.go#L177 instead. So please remove the `GetClientID()` method from the scheme interface.

File: code/go/0chain.net/chaincore/transaction/entity.go [hunk 2]
```diff
@@ -314,6 +315,7 @@ func (t *Transaction) GetSignatureScheme(ctx context.Context) (encryption.Signat
 	if err != nil {
 		co = client.NewClient()
 		co.ID = t.ClientID
+
 		if err := co.SetPublicKey(t.PublicKey); err != nil {
 			return nil, err
 		}
```
Comment (peterlimg) [inline code/go/0chain.net/chaincore/transaction/entity.go]: We should also check the ClientID and public key here before put it to cache.

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/chaincore/transaction/entity.go (hunk 2 above)
Comment (cnlangzi) [reply code/go/0chain.net/chaincore/transaction/entity.go]: fixed

File: code/go/0chain.net/core/encryption/bls0chain.go (hunk 1 above)
Comment (cnlangzi) [reply code/go/0chain.net/core/encryption/bls0chain.go]: fixed

File: code/go/0chain.net/chaincore/transaction/entity.go [hunk 3]
```diff
@@ -309,9 +310,14 @@ func (t *Transaction) VerifySignature(ctx context.Context) error {
...
 /*GetSignatureScheme - get the signature scheme associated with this transaction */
 func (t *Transaction) GetSignatureScheme(ctx context.Context) (encryption.SignatureScheme, error) {
-	var err error
+
 	co, err := client.GetClientFromCache(t.ClientID)
 	if err != nil {
+		err = encryption.VerifyPublicKeyClientID(t.PublicKey, t.ClientID)
+		if err != nil {
+			return nil, err
+		}
```
Comment (peterlimg) [inline code/go/0chain.net/chaincore/transaction/entity.go]: Hmm, I checked the code again, this ClientID and PublicKey should be verified when receiving the txn. I'll move the verification code there if we haven't done it yet.

**Review by peterlimg — APPROVED** [review]
LGTM

**Pull request merged**